[SPARK-50493][SPARK-37687][BUILD] Migrate kubernetes-client from `6.x` to `7.x`

panbingkun opened the pull request:
### What changes were proposed in this pull request?
The pr aims to migrate kubernetes-client from `6.x` to `7.x`, includes:
- upgrade `kubernetes-client` from `6.13.2` to `7.0.1`.
- switch default `HttpClient` implementation from `OkHttp` to `Vert.x`.


### Why are the changes needed?
- The full release notes: 
  https://github.com/fabric8io/kubernetes-client/releases/tag/v7.0.1
  https://github.com/fabric8io/kubernetes-client/releases/tag/v7.0.0
- [Vert.x as default HttpClient implementation](https://github.com/fabric8io/kubernetes-client/blob/v7.0.0/doc/MIGRATION-v7.md#vertx-as-default-httpclient-implementation-)

### Does this PR introduce _any_ user-facing change?
No.


### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

panbingkun committed:
[SPARK-50493][BUILD] Migrate kubernetes-client to 7.0.0

panbingkun committed:
fix

panbingkun committed:
fix

panbingkun committed:
setUseDaemonThread(true) for vert.x thread

panbingkun commented on dev/deps/spark-deps-hadoop-3-hive-2.3:
```diff
@@ -227,8 +228,8 @@ netty-transport-native-kqueue/4.1.115.Final/osx-x86_64/netty-transport-native-kq
 objenesis/3.3//objenesis-3.3.jar
-okhttp/3.12.12//okhttp-3.12.12.jar
-okio/1.17.6//okio-1.17.6.jar
+okhttp/3.14.2//okhttp-3.14.2.jar
```
After removing the mandatory specification of `okio` dependency, the version of `esdk-obs-java-3.20.4.2` dependency was restored as follows:
https://repo1.maven.org/maven2/com/huaweicloud/esdk-obs-java/3.20.4.2/esdk-obs-java-3.20.4.2.pom

panbingkun committed:
Merge branch 'master' into k8s_client_7_vert_x

panbingkun committed:
fix

panbingkun committed:
Merge branch 'master' into k8s_client_7_vert_x

panbingkun committed:
Merge branch 'master' into k8s_client_7_vert_x

panbingkun committed:
fix

panbingkun commented:
- After more than a `week` of investigation and research, when we upgraded `Kubernetes client` from `6.x` to `7.0` and switched `default` HttpClient implementation from `OkHttp` to `Vert.x`, the issue of `Kubernetes Integration test` failure has been resolved.
- The root cause is that some thread attributes `daemon` of `Vert.x` are set to `false`, so that after using `ProcessBuilder` to start the spark application on `K8S` (based on `spark-submit`), the `main thread` has already ended, but the above threads will not end, resulting in the overall `process` based on `spark-submit` not ending either.

panbingkun committed:
improve

panbingkun commented on resource-managers/kubernetes/core/src/main/java/org/apache/spark/deploy/k8s/SparkVertxHttpClientFactory.java:
```diff
@@ -0,0 +1,73 @@
+    this.vertx = createVertxInstance();
+  }
+
+  public static SparkVertxHttpClientFactory instance() {
```
why singleton?
https://github.com/fabric8io/kubernetes-client/issues/6709#issuecomment-2533614974

panbingkun commented:
cc @dongjoon-hyun @pan3793 @Yikun @LuciferYang

panbingkun committed:
fix

panbingkun committed:
Merge branch 'master' into k8s_client_7_vert_x

dongjoon-hyun reviewed (changes requested):
Please restore `okio` version to 1.17.6 explicitly.

It's because the downgrade is not supposed to be happen.
```
- okio/1.17.6//okio-1.17.6.jar
+ okio/1.17.2//okio-1.17.2.jar
```

panbingkun commented:
> Please restore `okio` version to 1.17.6 explicitly.
> 
> It's because the downgrade is not supposed to be happen.
> 
> ```
> - okio/1.17.6//okio-1.17.6.jar
> + okio/1.17.2//okio-1.17.2.jar
> ```

Okay.

panbingkun committed:
Merge branch 'master' into k8s_client_7_vert_x

dongjoon-hyun commented:
FYI, `okio` 1.7.16 was CVE-related updated, @panbingkun .
- https://github.com/apache/spark/pull/47758

cc @roczei , @yaooqinn from #47758

panbingkun committed:
okio 1.17.6

dongjoon-hyun commented:
+1 if that's the only one.

panbingkun commented on pom.xml:
```diff
@@ -235,8 +235,7 @@
     <!-- org.fusesource.leveldbjni will be used except on arm64 platform. -->
     <leveldbjni.group>org.fusesource.leveldbjni</leveldbjni.group>
-    <kubernetes-client.version>6.13.4</kubernetes-client.version>
-    <okio.version>1.17.6</okio.version>
```
Move to `hadoop-cloud/pom.xml`

panbingkun commented on pom.xml:
```diff
@@ -2594,11 +2593,6 @@
         <artifactId>javax.servlet-api</artifactId>
         <version>${javaxservlet.version}</version>
       </dependency>
-      <dependency>
```
Move to `hadoop-cloud/pom.xml`

pan3793 commented on resource-managers/kubernetes/core/src/main/java/org/apache/spark/deploy/k8s/SparkVertxHttpClientFactory.java:
```diff
@@ -0,0 +1,57 @@
+    try {
+      System.setProperty("vertx.disableDnsResolver", "true");
+      vertx = Vertx.vertx((new VertxOptions())
+        .setUseDaemonThread(true)
```
Non-daemon threads block app exit is a common issue for Spark on K8s, and this does not happen on YARN, which surprises a lot of users migrate their workloads from YARN to K8s, I think https://github.com/apache/spark/pull/46889 might help

panbingkun replied:
thanks!

panbingkun commented:
> +1 if that's the only one.

Yea, I use following command to analy it. Currently, that's the only one.
```shell
./build/mvn dependency:tree -Phadoop-3 -Phadoop-cloud -Phive-thriftserver -Pkinesis-asl -Pspark-ganglia-lgpl -Pkubernetes -Pvolcano -Pyarn -Pdocker-integration-tests -Pconnect -Phive -Pyarn -Pjvm-profiler > dependency/20241216.txt
```

panbingkun commented:
I think this PR can wait a little longer, because after https://github.com/fabric8io/kubernetes-client/pull/6726 is fixed, this PR can be simpler, see the following discussion: https://github.com/fabric8io/kubernetes-client/issues/6709#issuecomment-2545071491

dongjoon-hyun commented:
Thank you for the info, @panbingkun . Yes, let's wait.

panbingkun committed:
update to 7.0.1

panbingkun committed:
fix

panbingkun commented:
The version `7.0.1` has been released, and this PR has also been updated synchronously. 
It now looks much simpler.

panbingkun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala:
```diff
@@ -78,10 +74,6 @@ object SparkKubernetesClientFactory extends Logging {
       .getOption(s"$kubernetesAuthConfPrefix.$CLIENT_KEY_FILE_CONF_SUFFIX")
     val clientCertFile = sparkConf
       .getOption(s"$kubernetesAuthConfPrefix.$CLIENT_CERT_FILE_CONF_SUFFIX")
-    // TODO(SPARK-37687): clean up direct usage of OkHttpClient, see also:
```
After this pr, I think [Cleanup direct usage of OkHttpClient](https://issues.apache.org/jira/browse/SPARK-37687) can also be closed.

dongjoon-hyun replied:
I agree with you. I added SPARK-37687 to the PR title.

pan3793 reviewed (approved):
thanks, code change lgtm, NOTICE and LICENSE might need to be updated since it pulls new deps.

dongjoon-hyun commented:
Thank you for syncing with 7.0.1, @panbingkun .

dongjoon-hyun reviewed (changes requested):
+1 for the above comment about LICENSE .

In addition, to be safe, please keep the original `okhttp` version by overriding the transitive dependency, @panbingkun . We had better handle `okhttp` upgrade independently, too.
```
- okhttp/3.12.12//okhttp-3.12.12.jar
+ okhttp/3.14.2//okhttp-3.14.2.jar
```

panbingkun commented:
Okay, I'll update it.

panbingkun committed:
Merge branch 'master' into k8s_client_7_vert_x

panbingkun committed:
update

panbingkun commented:
The `LICENSE` of the newly added 4 jars are as follows:

|Jar|LICENSE|URL|
|----|----|--------------------|
|io.vertx:vertx-auth-common|Apache License Version 2.0|https://github.com/eclipse-vertx/vertx-auth/blob/master/LICENSE.txt|
|io.vertx:vertx-core|Eclipse Public License 2.0 or Apache Software License 2.0|https://github.com/eclipse-vertx/vert.x/blob/master/LICENSE.md|
|io.vertx:vertx-web-client|Apache License Version 2.0|https://github.com/vert-x3/vertx-web/blob/master/LICENSE.txt|
|io.vertx:vertx-web-common|Apache License Version 2.0|https://github.com/vert-x3/vertx-web/blob/master/LICENSE.txt|

panbingkun commented:
> +1 for the above comment about LICENSE .
> 
> In addition, to be safe, please keep the original `okhttp` version by overriding the transitive dependency, @panbingkun . We had better handle `okhttp` upgrade independently, too.
> 
> ```
> - okhttp/3.12.12//okhttp-3.12.12.jar
> + okhttp/3.14.2//okhttp-3.14.2.jar
> ```

Done.
----
But I'm not sure if the `LICENSE` I added for these 4 newly added jars is correct. 
I have added the original `LICENSE` of the 4 newly added jars in the comments for everyone to review.
@pan3793 @dongjoon-hyun Can you help me take a look at it?

panbingkun committed:
update

pan3793 commented on LICENSE-binary:
```diff
@@ -515,6 +518,7 @@ org.glassfish.jaxb:jaxb-runtime
 
 Eclipse Public License (EPL) 2.0
 --------------------------------
+io.vertx:vertx-core https://github.com/eclipse-vertx/vert.x
```
for dual-licensed 3rd deps, we only need to pick one. for this case, it's under EPL-2.0 and AL 2.0, simply pick the AL 2.0 and no additional LICENSE copy is required.

panbingkun replied:
Because these are `dual-licensed 3rd deps`, should we choose `EPL-2.0` or `AL 2.0`?
Is there a `clear standard` for this? Is it still possible to `freely express oneself`?

pan3793 replied:
the downstream projects are free to pick one or all licenses they want, as an ASF project, obviously, we prefer to choose AL 2.0

LuciferYang replied:
> the downstream projects are free to pick one or all licenses they want, as an ASF project, obviously, we prefer to choose AL 2.0

If so, +1

dongjoon-hyun replied:
+1 for AL 2.0, too.

panbingkun replied:
Updated, thanks all!

dongjoon-hyun commented:
BTW, sorry for the late response, @pan3793 .

panbingkun committed:
Merge branch 'master' into k8s_client_7_vert_x

panbingkun committed:
update LICENSE

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @panbingkun, @LuciferYang , and @pan3793 .

Merged to master for Apache Spark 4.0.0. Happy New Year!

dongjoon-hyun commented:
I collected this as a subtask for the umbrella JIRA, SPARK-49524 (Improve K8s support).

attilapiros commented:
> > Please restore `okio` version to 1.17.6 explicitly.
> > It's because the downgrade is not supposed to be happen.
> > ```
> > - okio/1.17.6//okio-1.17.6.jar
> > + okio/1.17.2//okio-1.17.2.jar
> > ```
> 
> Okay.
> 
> Because after switch default HttpClient implementation from `OkHttp` to `Vert.x`, `kubernetes-client` no longer needs to rely on `okio`. Currently, only `hadoop-cloud-storage` depends on it (`okio`), as follows:
> 
> ```
> [INFO] +- org.apache.hadoop:hadoop-cloud-storage:jar:3.4.1:compile
> [INFO] |  +- org.apache.hadoop:hadoop-annotations:jar:3.4.1:compile
> [INFO] |  +- org.apache.hadoop:hadoop-aliyun:jar:3.4.1:compile
> [INFO] |  |  +- com.aliyun.oss:aliyun-sdk-oss:jar:3.13.2:compile
> [INFO] |  |  |  +- org.jdom:jdom2:jar:2.0.6:compile
> [INFO] |  |  |  +- com.aliyun:aliyun-java-sdk-core:jar:4.5.10:compile
> [INFO] |  |  |  |  +- org.ini4j:ini4j:jar:0.5.4:compile
> [INFO] |  |  |  |  +- io.opentracing:opentracing-api:jar:0.33.0:compile
> [INFO] |  |  |  |  \- io.opentracing:opentracing-util:jar:0.33.0:compile
> [INFO] |  |  |  |     \- io.opentracing:opentracing-noop:jar:0.33.0:compile
> [INFO] |  |  |  +- com.aliyun:aliyun-java-sdk-ram:jar:3.1.0:compile
> [INFO] |  |  |  \- com.aliyun:aliyun-java-sdk-kms:jar:2.11.0:compile
> [INFO] |  |  \- org.codehaus.jettison:jettison:jar:1.5.4:compile
> [INFO] |  +- org.apache.hadoop:hadoop-azure-datalake:jar:3.4.1:compile
> [INFO] |  |  \- com.microsoft.azure:azure-data-lake-store-sdk:jar:2.3.9:compile
> [INFO] |  \- org.apache.hadoop:hadoop-huaweicloud:jar:3.4.1:compile
> [INFO] |     \- com.huaweicloud:esdk-obs-java:jar:3.20.4.2:compile
> [INFO] |        +- com.jamesmurty.utils:java-xmlbuilder:jar:1.2:compile
> [INFO] |        +- com.squareup.okhttp3:okhttp:jar:3.14.2:compile
> [INFO] |        \- com.squareup.okio:okio:jar:1.17.6:compile
> ```
> 
> So I will migrate the dependency declaration for `com.squareup.okio:okio` from `pom.xml` to `hadoop-cloud/pom.xml` to reduce misunderstandings, WDYT? @dongjoon-hyun


Do I get this right: `okio` and `okhttp` are only needed because of `huaweicloud` which is not part of upstream spark?
